fs-index/src/index.rs

```diff
@@ -474,4 +474,109 @@ impl<Id: ResourceId> ResourceIndex<Id> {
 
         Ok(IndexUpdate { added, removed })
     }
+
+    /// Track the addition of a new resource
+    ///
+    /// This method accepts a resource ID and returns `Result::Ok` if the
+    /// resource was successfully added to the index.
+    ///
+    /// **Note**: The called must ensure that:
+    /// - The index is up-to-date with the file system except for the new
+    ///   resource
+    /// - The resource is not already in the index
+    pub fn track_addition(&mut self, _id: Id) -> Result<()> {
+        /*
+           # Given:
+            - The file system contains resources with duplicate IDs
+              (e.g., hash collisions or files with the same content)
+            - The index is up-to-date with the file system except for the new
+              resource
+            - The resource is not already in the index
+
+           # Problem:
+           - We only know the ID of the resource
+           - We need to get the path of the resource so to add it to the index
+           - To get the path, we need to scan the file system for the resource
+             that fulfills theses conditions:
+                - The resource has the given ID
+                - The resource path is not in the index
+            To do this, we need to iterate over all the resources in the file
+            system and check if the resource has the given ID and is not in the
+            index (complexity: O(n)).
+           - Track API should be O(1) or O(log n) if possible because it is
+             used for fine-grained control over tracking changes in the index
+        */
+
+        unimplemented!("Track addition is not implemented yet");
+    }
+
+    /// Track the removal of a resource
+    ///
+    /// This method accepts a resource ID and returns `Result::Ok` if the
+    /// resource was successfully removed from the index.
+    ///
+    /// **Note**: The called must ensure that:
+    /// - The index is up-to-date with the file system except for the removed
+    ///  resource
+    /// - The resource is already in the index
+    pub fn track_removal(&mut self, _id: Id) -> Result<()> {
```

Comment:
We can consider enforcing invariant using `debug_assert!` that the id is present in the index. If we do so, we also can remove `Result` and either succeed or panic if the invariant is violated.

Reply:
Consider merging with `update_one` when we pass the removed path and just remove it from the index together with the current id.

Reply:
> We can consider enforcing invariant using `debug_assert!` that the id is present in the index.

If we decide to accept the path as a parameter in `track_addition()` as mentioned [here](https://github.com/ARK-Builders/ark-core/pull/85#discussion_r1718357077), it would be more consistent to also accept the path instead of the resource ID in `track_deletion()`.

That said, I believe [having a single function to update the index for one resource](https://github.com/ARK-Builders/ark-core/pull/85#discussion_r1719841869) might be more effective than separate functions for `track_addition()` and `track_deletion()`. What are your thoughts?